sdks/java/extensions/sketching/src/main/java/org/apache/beam/sdk/extensions/sketching/frequency/KMostFrequent.java

```diff
@@ -0,0 +1,200 @@
+   * distinct key in the input {@code PCollection} to a sketch which contains the most frequent
+   * values associated with that key in the input {@code PCollection}.
+   *
+   * <p>The {@code capacity} parameter controls the maximum number of elements the sketch
+   * can contain. Once this capacity is reached, the least frequent element is dropped each
+   * time an incoming element is not already present in the sketch.
+   * Each element in the sketch is associated to a counter, that keeps track of the estimate
+   * frequency as well as the maximal potential error.
+   * <br>See {@link KMostFrequentFn} for more details.
+   *
+   * <p>Example of use
+   * <pre> {@code PCollection<KV<Integer, String>> input = ...;
+   * PCollection<KV<Integer, StreamSummary<String>>> ssSketch = input
+   *        .apply(SketchFrequencies.<Integer, String>globally(10000));
+   * } </pre>
+   *
+   * @param capacity    the maximum number off distinct elements that the Stream Summary can keep
+   *                    track of at the same time
+   * @param <K>         the type of the keys in the input and output {@code PCollection}s
+   * @param <InputT>    the type of values in the input {@code PCollection}
+   */
+  public static <K, InputT> Combine.PerKey<K, InputT, StreamSummary<InputT>> perKey(int capacity) {
+    if (capacity < 1) {
+        throw new IllegalArgumentException("The capacity must be strictly positive");
+    }
+    return Combine.<K, InputT, StreamSummary<InputT>>perKey(new KMostFrequentFn<InputT>(capacity));
+  }
+
+  /**
+   * A {@code Combine.CombineFn} that computes the stream into a {@link StreamSummary}
+   * sketch, useful as an argument to {@link Combine#globally} or {@link Combine#perKey}.
+   *
+   * <p>The Space-Saving algorithm summarizes the stream by using a doubly linked-list of buckets
+   * ordered by the frequency value they represent. Each of these buckets contains a linked-list
+   * of counters, which estimate the {@code count} for an element as well as the maximum
+   * overestimation {@code e} associated to it. The frequency cannot be overestimated.
+   *
+   * <p>An element is guaranteed to be in the top K most frequent if its guaranteed number of hits,
+   * i.e. {@code count - e}, is greater than the count of the element at the position k+1.
+   *
+   * @param <InputT> the type of the elements being combined
+   */
+  static class KMostFrequentFn<InputT>
+          extends Combine.CombineFn<InputT, StreamSummary<InputT>, StreamSummary<InputT>> {
+
+    private int capacity;
+
+    private KMostFrequentFn(int capacity) {
+      this.capacity = capacity;
+    }
+
+    @Override
+      public StreamSummary<InputT> createAccumulator() {
+      return new StreamSummary<>(this.capacity);
+    }
+
+    @Override
+    public StreamSummary<InputT> addInput(StreamSummary<InputT> accumulator, InputT element) {
+      accumulator.offer(element, 1);
```

Comment:
<!--new_thread; commit:6bf51fdf6afe6f389415792f335dd08adf9fa85a; resolved:0-->
element may not implement a correct equals, for example byte[].

You'll want to create a wrapper types that implement Externalizable which stores a reference to the element, the element coder, and the structural value for the element. This wrapper type should delegate equals to the structural value and its write/read method should use the coder and element to encode/decode the wrapper with the structural value being ephemeral.

Please add a test for this case by counting 1000 new instances of byte[]{ 0x00 } or something similar.